File: 329-longest-increasing-path-in-a-matrix/longest-increasing-path-in-a-matrix.py
class Solution:
    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        n = len(matrix)
        m = len(matrix[0])
        dx = [0,0,-1,1]
        dy = [1,-1,0,0]
        res = 0
        
        @cache
        def dfs(x, y):
            res = 0
            for i in range(4):
                a = x + dx[i]
                b = y + dy[i]
                if a < 0 or a >= n or b < 0 or b >= m: continue
                if matrix[a][b] <= matrix[x][y]: continue
                res = max(res, dfs(a,b))

            res+=1
            return res

        for i in range(n):
            for j in range(m):
                res = max(res, dfs(i,j))
        
        return res

File: 329-longest-increasing-path-in-a-matrix/longest-increasing-path-in-a-matrix.cpp
class Solution {
public:
    vector<vector<int>> f, w;
    int n,m;
    int dx[4] = {0,0,-1,1};
    int dy[4] = {1,-1,0,0};
    int longestIncreasingPath(vector<vector<int>>& matrix) {
        if (matrix.empty() || matrix[0].empty()) return 0;
        n = matrix.size(), m = matrix[0].size();
        w = matrix;
        f = vector(n, vector<int> (m, -1));
        int res = 0;
        for (int i = 0; i < n; i++){
            for (int j = 0; j < m; j++){
                res = max(res, dp(i,j));
            }
        }
        return res;
    }

    int dp(int x, int y) {
        auto &v = f[x][y];
        if (v != -1) return v;

        v = 1;

        for (int i = 0; i < 4; i++){
            int a = x + dx[i];
            int b = y + dy[i];
            if (a < 0 || a >= n || b < 0 || b >= m) continue;
            if (w[a][b] > w[x][y]) {
                v = max(v, dp(a,b)+1);
            }
        }
        return v;
        
    }

};